rust/lance/src/io/exec/scalar_index.rs

```diff
@@ -257,29 +257,76 @@ impl MaterializeIndexExec {
     ) -> Result<RecordBatch> {
         // TODO: multiple batches, stream without materializing all row ids in memory
         let mask = expr.evaluate(dataset.as_ref());
-        let fragment_bitmap = RoaringBitmap::from_iter(fragments.iter().map(|frag| frag.id as u32));
-        // The user-requested `fragments` is guaranteed to be stricter than the index's fragment
-        // bitmap.  This node only runs on indexed fragments and any fragments that were deleted
-        // when the index was trained will still be deleted when the index is queried.
-        let prefilter = PreFilter::create_deletion_mask(dataset.clone(), fragment_bitmap);
+        let span = debug_span!("create_prefilter");
+        let prefilter = span.in_scope(|| {
+            let fragment_bitmap =
+                RoaringBitmap::from_iter(fragments.iter().map(|frag| frag.id as u32));
+            // The user-requested `fragments` is guaranteed to be stricter than the index's fragment
+            // bitmap.  This node only runs on indexed fragments and any fragments that were deleted
+            // when the index was trained will still be deleted when the index is queried.
+            PreFilter::create_deletion_mask(dataset.clone(), fragment_bitmap)
+        });
         let mask = if let Some(prefilter) = prefilter {
             let (mask, prefilter) = futures::try_join!(mask, prefilter)?;
             mask.also_block((*prefilter).clone())
         } else {
             mask.await?
         };
-        let ids = match (mask.allow_list, mask.block_list) {
+        let span = debug_span!("make_ids");
+        let ids = span.in_scope(|| match (mask.allow_list, mask.block_list) {
             (None, None) => FragIdIter::new(fragments).collect::<Vec<_>>(),
-            (Some(allow_list), None) => FragIdIter::new(fragments)
-                .filter(|row_id| allow_list.contains(*row_id))
-                .collect(),
+            (Some(allow_list), None) => {
+                let allowed_ids = allow_list.row_ids();
+                if let Some(allowed_ids) = allowed_ids {
+                    let frag_id_filter = fragments
+                        .iter()
+                        .map(|frag| frag.id as u32)
+                        .collect::<HashSet<_>>();
+                    allowed_ids
+                        .filter_map(|addr| {
+                            if frag_id_filter.contains(&addr.fragment_id()) {
+                                Some(u64::from(addr))
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>()
```

Comment:
This seems suboptimal. `allow_list` is already a map from `fragment_id` to local row ids. Maybe it would make more sense to remove the fragments at that stage, and then afterward turn it into a vector of row addressed?

Something like:

```rust
for frag in fragments {
    allow_list.remove_fragment(frag.id);
}
allow_list.row_ids().collect::<Vec<_>>()
```

Reply:
:facepalm: you're absolutely right.

Reply:
Fixed.  I added `remove_fragments`.